source/updateCheck.py

```diff
@@ -352,7 +354,8 @@ def __init__(self, parent, destPath, version):
 		self.Sizer = mainSizer
 		mainSizer.Fit(self)
 		self.Center(wx.BOTH | wx.CENTER_ON_SCREEN)
-		self.Show()
+		gui.mainFrame.prePopup()

```

Comment:
what's this about?

Reply:
If you call this, I think you also need to call `postPopup()`.

Reply:
Yes, you'd think so. The use of prePopup and postPopup inside updateCheck really puzzles me.

1. prePopup is only used in the UpdateDownloader, not for the other dialogs. It was called for the old update installation message box, so that's why I belief it should be called in the UpdateAskInstallDialog as well.
2. postPopup is even more confusing given it is being called in a wx.CallLater with a not very descriptive comment. A git blame points at @jcsteh, so may be he knows more?
3. In several places in the gui module, you see the following code:

    ```
mainFrame.prePopup()
dialog.Show()
mainFrame.postPopup()
```

    However, this doesn't seem to make much sense, as Show doesn't block the caller. For ShowModal, it does make sense as ShowModal blocks the caller until the dialog is dismissed.

Reply:
Perhaps its better to use `def runScriptModalDialog` in `source/gui/__init__`

Reply:
TL;DR: The general rule is that if you're popping up a window (as opposed to a window which appears as a result of some action performed while our GUI was already visible), you must call `prePopup`. If you call `prePopup`, there must be a corresponding call to `postPopup`. Ideally, this `postPopup` call would occur once the window is dismissed, but for windows that aren't modal, it's enough in most cases to just call it after `Show`.

`prePopup` and `postPopup` have two major functions:

1. To force Windows to bring us to the foreground (`prePopup`) and to return to the previous foreground (`postPopup`). The way foreground stuff works in Windows is really obscure. `prePopup` mostly makes sense; we ask Windows to bring us to the foreground. However, we were seeing cases where the user would just be stuck in limbo when our pop-ups were dismissed. I still don't really understand why this is the case, but the show/hide in `postPopup` seemed to fix it. (Obviously, that doesn't actually get applied at the right tie for dialogs that aren't modal, but it seems to behave regardless.) This code is very old, wx has been updated and Windows XP is no longer a factor, so you could consider re-assessing to see whether we can get rid of this ugliness. However, note that installed versus portable is still a factor, since uiAccess gives the app special privileges regarding the foreground window.
2. These methods also set things up so that NVDA is aware of what the focus was before the pop-up. This is important for input gestures, for example, so it knows which object to get gestures for. `postPopup` does the necessary teardown of this state. The fact that it gets called immediately after `Show` is fine in this case because the dialog should already have done what it needs to do with this state.